src/v/storage/log_manager.cc

```diff
@@ -401,6 +402,71 @@ ss::future<> log_manager::remove(model::ntp ntp) {
     });
 }
 
+ss::future<> log_manager::remove_orphan(
+  ss::sstring data_directory_path, model::ntp ntp, model::revision_id rev) {
+    vlog(stlog.info, "Asked to remove orphan for: {} revision: {}", ntp, rev);
+
+    auto topic_directory_path = (std::filesystem::path(data_directory_path)
+                                 / ntp.topic_path())
+                                  .string();
+
+    if (_logs.contains(ntp)) {
+        co_return;
+    }
+    auto topic_directory_exist = co_await ss::file_exists(topic_directory_path);
+    if (!topic_directory_exist) {
+        co_return;
+    }
+
+    try {
+        auto topic_directory = co_await ss::open_directory(
+          topic_directory_path);
+        co_await topic_directory
+          .list_directory(
+            [&ntp, &topic_directory_path, &rev](ss::directory_entry entry) {
+                auto ntp_directory_data
+                  = ntp_directory_path::parse_partition_directory(entry.name);
+                if (!ntp_directory_data) {
+                    return ss::now();
+                }
+                if (
+                  ntp_directory_data->partition_id == ntp.tp.partition
+                  && ntp_directory_data->revision_id <= rev) {
+                    auto ntp_directory = std::filesystem::path(
+                                           topic_directory_path)
+                                         / std::filesystem::path(entry.name);
+                    vlog(
+                      stlog.info,
+                      "Cleaning up ntp [{}] rev {} directory {} ",
+                      ntp,
+                      ntp_directory_data->revision_id,
+                      ntp_directory);
+                    return ss::recursive_remove_directory(ntp_directory);
+                }
+                return ss::now();
+            })
+          .done()
+          .finally(
+            [topic_directory]() mutable { return topic_directory.close(); });
+    } catch (ss::broken_promise const&) {
```

Comment:
we should not catch `broken_promise` exception as it indicates some other issue f.e. promise is deleted before setting the result,

Reply:
It occurs when we delete directory topic directory in other shard before closing it here.
We don't have any cross-shard mutex to do it, so I came up with this solution

Reply:
This smells like a bug in the list_directory() implementation in seastar. It seems like if there is an exception, the underlying stream is destroyed without calling close destroying the promise.